primes/matho-pascal.1
.TH MATHO-PASCAL 1 "" "Mathomatic" "Mathomatic Utilities"

.SH NAME
matho-pascal \- display Pascal's triangle

.SH SYNOPSIS
.B matho-pascal
[number-of-lines]

.SH DESCRIPTION
This command-line utility is optionally part of the
.BR mathomatic (1)
package.
It calculates up to 1000 lines of Pascal's triangle using floating point arithmetic,
dumping the lines to standard output.
The default is to center one screen full.

Every number inside Pascal's triangle is the sum of the two numbers
immediately above it.

Each line of Pascal's triangle is the same as the binomial coefficients
for a given power.

The sum of all numbers in each line of Pascal's triangle is a power of 2.

.SH AUTHOR 
George Gesslein II (gesslein@mathomatic.org)
at "http://www.mathomatic.org".

.SH "REPORTING BUGS"
If you find a bug, please report it to the author
or at "https://launchpad.net/mathomatic".

.SH "SEE ALSO"
.BR mathomatic (1),
.BR matho-primes (1),
.BR matho-sumsq (1)


NEWS

The latest version of Mathomatic is 16.0.5, here are the changes:



CHANGES MADE TO MATHOMATIC 16.0.4 TO BRING IT UP TO THE NEXT VERSION:

09/17/12 - Corrected and improved "tests/finance.in" and the terminology used,
           thanks to Wolfram Alpha.

09/20/12 - Fixed "./update" command (which updates "proto.h") to work without
           the readline development headers and libraries installed.

09/23/12 - m4/gradians.m4 copied with m4/functions.m4, when installing with
           "make m4install" into "/usr/local/share/mathomatic/m4/",
           so gradian trig mode can be used by typing "rmath gradians.m4"
           at the shell prompt, while in that directory.
           The debug level was being set in degrees.m4
           and gradians.m4; no longer.

09/25/12 - Mathomatic now always runs "set load" after "set save",
           for the convenience of immediately loading the specified settings,
           and for error checking.

           From now on Mathomatic does not set the alarm nor hangup handlers,
           because it crashes the entire shell window when readline
           is compiled in and an alarm or hangup signal happens.
           Works acceptably now that these signal handlers are not set,
           it just quits when time is up, saying "Alarm clock";
           The same thing happens for SIGHUP, except it just says "Hangup"
           and quits.  I am really not liking readline nor its license now.
           SIGALRM and SIGHUP are no longer given handlers.  The default
           seems to be what has to be there with the current readline mess.
           Overwriting the current signal handlers causes serious bugs,
           but only for readline.

09/29/12 - Fixed "lib/compile.testmain" to work with the latest linker.

10/16/12 - Made all output methods respect "set columns", so
           "display all >output.txt" will not always use infinite columns.
           This goes for the Symbolic Math Library, too.

Mathomatic version 16.0.5 released Sunday 10/21/12.

CHANGES MADE TO MATHOMATIC 16.0.3 TO BRING IT UP TO THE NEXT VERSION:

09/09/12 - Allow use of editline library in Mathomatic, because somehow
           use of GPL libraries in LGPL code is not allowed, and GNU readline
           is GPL.  I am so confused about this Debian bug#687063:
           http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=687063

           Based on the bug report's information sources,
           it appears he is saying the truth about this license issue,
           so I will have to upload new versions of
           Mathomatic with editline instead of readline to Debian.
           The only noticeable difference should be it doesn't save
           the history between Mathomatic sessions.
           I will comply with all that request it,
           though most distributions do not include editline,
           which has no major licensing restrictions.  Readline is still
           perfectly usable and good as before, you just have to link
           it in yourself.

09/10/12 - Going to have to make a new release already, 2 days after the last
           one, so I can upload this readline licensing fix to Debian by
           linking with editline.  There is no reason for anyone to upgrade
           to version 16.0.4, unless you wish to link with editline instead
           of readline.  The proper code has been added.  All you have to do
           is have the editline libraries loaded on your system,
           and run "make EDITLINE=1" to compile and link Mathomatic with
           editline.

Mathomatic version 16.0.4 released Monday 09/10/12.

CHANGES MADE TO MATHOMATIC 16.0.2 TO BRING IT UP TO THE NEXT VERSION:

New command "set load" loads the current set options startup file again,
displaying the startup file as it reads it in.
If the file doesn't exist, or something is wrong, then an error message is
displayed and the set command returns with failure.
Accidently putting "load" in the startup file is now handled correctly.

08/09/12 - Allow "matho-primes all" and the command "list primes all" in
           Mathomatic to continually output consecutive prime numbers.

08/12/12 - Fixed any rman errors in the makefile so they won't be ignored.

08/17/12 - rmath and matho no longer set the debug_level or modulus_mode,
           so they can be set by the startup options file.

08/19/12 - Allow the repeat prefix on the approximate command, making it
           approximate and simplify as much as the calculate command
           does.  Hopefully someday it will just give a temporary
           result, too.  Numerical input into the symbolic math library
           now uses "repeat approximate" instead of just "approximate"
           to approximate the user's input, so the result is the
           same as the Mathomatic application's result.

           Disallow the same warnings to be repeatedly displayed, if the
           current warning is the same as the previous warning.

08/20/12 - Allow directory names as read command arguments, instructing
           the read command to change the current directory to the specified
           directory.  Without any arguments, the read command now does an
           "ls" command in Unix/Linux, and "dir" under MS-Windows, listing
           the current directory contents.  Running Mathomatic with a
           directory name argument now conveniently changes directory to
           that directory, then gives you the main prompt.

08/28/12 - Made "integer" type variables much more useful.  For example,
           the following now happens generally:

1-> i^(4*integer)
#1: i^(4*integer)
1-> simplify
#1: 1
1-> i^((4*integer) + 1)
#2: i^((4*integer) + 1)
2-> simplify
#2: i
2-> i^((4*integer) + 2)
#3: i^((4*integer) + 2)
3-> simplify
#3: -1
3-> i^((4*integer) + 3)
#4: i^((4*integer) + 3)
4-> simplify
#4: -1*i

08/29/12 - Displays "Calculating..." whenever autocalc is used now.

08/30/12 - Added ability to set the normal text color.

primes/matho-pascal.c
/*
 * Calculate and display Pascal's triangle.
 *
 * Copyright (C) 2005 George Gesslein II.
 
This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

The chief copyright holder can be contacted at gesslein@mathomatic.org, or
George Gesslein II, P.O. Box 224, Lansing, NY  14882-0224  USA.
 
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <string.h>
#include <errno.h>
#include <math.h>
#include <float.h>
#include <assert.h>
#if	!MINGW
#include <sys/ioctl.h>
#include <termios.h>
#endif

#define	true	1
#define	false	0

#ifndef	LDBL_DIG
#define	USE_DOUBLES	1
#warning "long doubles not supported by this C compiler."
#endif

#if	USE_DOUBLES
typedef double double_type;
#warning "Using only double floats instead of long double floats, as requested."
#else
typedef long double double_type;
#endif

#define	MAX_LINES	1000	/* max number of lines of Pascal's triangle allowed */

void	allocate_triangle(void);
void	calculate_triangle(void);
void	display_triangle(void);
int	center_buf(int line_no, int cell_size);
void	usage(int ev);

#if	USE_DOUBLES
int		precision = DBL_DIG;
#else
int		precision = LDBL_DIG;
#endif
int		glines = 26;
int		gcell_size = 6;
double_type	*garray[MAX_LINES];
int		screen_columns = 80;
int		centered = true;
char		gline_buf[1000];

char		*prog_name = "matho-pascal";

int
main(int argc, char *argv[])
{
#if	!MINGW
	struct winsize	ws;

	ws.ws_col = 0;
	ws.ws_row = 0;
	ioctl(1, TIOCGWINSZ, &ws);
	if (ws.ws_col > 0) {
		screen_columns = ws.ws_col;
	}
#endif
	if (screen_columns >= sizeof(gline_buf)) {
		screen_columns = sizeof(gline_buf) - 1;
	}

	switch (argc) {
	case 0:
	case 1:
		break;
	case 2:
		centered = false;
		if (isdigit(argv[1][0])) {
			glines = atoi(argv[1]);
			break;
		}
	default:
		usage(EXIT_FAILURE);
	}
	if (glines <= 0 || glines > MAX_LINES) {
		fprintf(stderr, "%s: Number of lines out of range (1..%d).\n", prog_name, MAX_LINES);
		exit(EXIT_FAILURE);
	}
	allocate_triangle();
	calculate_triangle();
	display_triangle();
	exit(EXIT_SUCCESS);
}

void
allocate_triangle(void)
{
	int	i;

	for (i = 0; i < glines; i++) {
		garray[i] = (double_type *) calloc(i + 1, sizeof(double_type));
		if (garray[i] == NULL) {
			fprintf(stderr, "%s: Not enough memory.\n", prog_name);
			exit(EXIT_FAILURE);
		}
	}
}

void
calculate_triangle(void)
{
	int	i, j;

	for (i = 0; i < glines; i++) {
		for (j = 0; j <= i; j++) {
			if (j == 0 || j == i) {
				garray[i][j] = 1.0;	/* the line begins and ends with 1 */
			} else {
				garray[i][j] = garray[i-1][j-1] + garray[i-1][j];
			}
		}
	}
}

void
display_triangle(void)
{
	int		i, j;
	int		len;

	if (centered && glines > 20) {
		len = center_buf(19, 8);
		if (len > 0 && len < screen_columns) {
			gcell_size = 8;	/* for very wide screens */
		}
	}
	for (i = 0; i < glines; i++) {
		if (centered) {
			len = center_buf(i, gcell_size);
			if (len <= 0 || len >= screen_columns) {
				return;	/* stop here because of wrap-around */
			}
			/* center on screen */
			for (j = (screen_columns - len) / 2; j > 0; j--) {
				printf(" ");
			}
			printf("%s", gline_buf);
		} else {
			for (j = 0; j <= i; j++) {
#if	USE_DOUBLES
				printf("%.*g ", precision, garray[i][j]);
#else
				printf("%.*Lg ", precision, garray[i][j]);
#endif
			}
		}
		printf("\n");
	}
}

/*
 * Create a line of output in gline_buf[] for centering mode.

 * Return length if successful, otherwise return 0.
 */
int
center_buf(int line_no, int cell_size)
{
	int	j, k;
	int	i1;
	int	len;
	char	buf2[100];

	assert(line_no < glines);
	gline_buf[0] = '\0';
	for (j = 0; j <= line_no; j++) {
#if	USE_DOUBLES
		len = snprintf(buf2, sizeof(buf2), "%.*g", precision, garray[line_no][j]);
#else
		len = snprintf(buf2, sizeof(buf2), "%.*Lg", precision, garray[line_no][j]);
#endif
		assert(len > 0);
		if (len >= cell_size) {
			return(0);	/* cell_size too small */
		}
		/* center in the cell */
		for (k = i1 = (cell_size - len) / 2; k > 0; k--) {
			strcat(gline_buf, " ");
		}
		strcat(gline_buf, buf2);
		for (k = len + i1; k < cell_size; k++) {
			strcat(gline_buf, " ");
		}
	}
	return(strlen(gline_buf));
}

void
usage(int ev)
{
	printf("Usage: %s [number-of-lines]\n\n", prog_name);
	printf("Display up to %d lines of Pascal's triangle.\n", MAX_LINES);
	printf("If number-of-lines is specified, don't center output.\n");
	printf("Number of digits of precision is %d.\n", precision);
	exit(ev);
}
